File: Leetcode-Python/IntersectionOfTwoLinkedLists.py
# encoding: utf-8
"""
Write a program to find the node at which the intersection of two singly linked lists begins.


For example, the following two linked lists:

A:          a1 → a2
                   ↘
                     c1 → c2 → c3
                   ↗
B:     b1 → b2 → b3
begin to intersect at node c1.


Notes:

If the two linked lists have no intersection at all, return null.
The linked lists must retain their original structure after the function returns.
You may assume there are no cycles anywhere in the entire linked structure.
Your code should preferably run in O(n) time and use only O(1) memory.
"""


class Solution:
    # @param two ListNodes
    # @return the intersected ListNode
    def getIntersectionNode(self, headA, headB):
        lenA = lenB = 0
        curA = headA
        curB = headB
        while curA or curB:
            if curA:
                lenA += 1
                curA = curA.next
            if curB:
                lenB += 1
                curB = curB.next

        while lenA > lenB:
            headA = headA.next
            lenA -= 1
        while lenA < lenB:
            headB = headB.next
            lenB -= 1

        while headA:
            if headA == headB:
                return headA
            headA = headA.next
            headB = headB.next
        return None

File: Java-By-Tag/src/main/IntersectionOfTwoLinkedLists.java
package src.main;

/**
 * Created by yhf on 8/27/15.
 */

/**
 * Write a program to find the node at which the intersection of two singly linked lists begins.
 *
 * For example, the following two linked lists:
 *
 * A:      a1 → a2
 *                ↘
 *                 c1 → c2 → c3
 *                ↗
 * B:  b1 → b2 → b3
 *
 * begin to intersect at node c1.
 *
 * Notes:
 *
 * If the two linked lists have no intersection at all, return null.
 * The linked lists must retain their original structure after the function returns.
 * You may assume there are no cycles anywhere in the entire linked structure.
 * Your code should preferably run in O(n) time and use only O(1) memory.
 */

public class IntersectionOfTwoLinkedLists {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        int lenA = getLength(headA), lenB = getLength(headB);
        int lenDiff = lenA - lenB;
        while (lenDiff < 0) {
            headB = headB.next;
            lenDiff++;
        }
        while (lenDiff > 0) {
            headA = headA.next;
            lenDiff--;
        }

        while (headA != null && headA != headB) {
            headA = headA.next;
            headB = headB.next;
        }
        return headA;
    }

    private int getLength(ListNode head) {
        int len = 0;
        while (head != null) {
            len++;
            head = head.next;
        }
        return len;
    }
}
